### ISSUE TITLE
Image pull starts to fail with DNS errors (i/o timeout)

### ISSUE THREAD

**Issue by jgoeres** [kubernetes/minikube#10280]
Hi,

I am currently observing a severe problem with my minikube environment. Running on Windows 10 with Hyper-V.
After "minikube start", as a first step I usually enable the ingress add-on.
Next, I start deploying a few workloads, most of which take images from our internal registry.

This seemed to work fine for a while, but now I am seeming randomly getting pods in "ErrImagePull" and eventually "ImagePullBackoff" state. 
Most of the time this starts when I deploy the main workload (~10 pods, all images from our internal registry), but I sometimes also observed it for the images of the ingress add-on (but never for the pods of the core minikube infrastructure).
Funny enough, most of the time for a few pods the image pull works and they reach "Running" state, while other go to (and stay in) "ImagePullBackoff", e.g.

    NAMESPACE       NAME                                        READY   STATUS                  RESTARTS   AGE
    kube-system     coredns-6955765f44-tbgrg                    1/1     Running                 0          32m
    kube-system     etcd-minikube                               1/1     Running                 0          31m
    kube-system     ingress-nginx-admission-create-dgl4c        0/1     Completed               0          26m
    kube-system     ingress-nginx-admission-patch-z9p96         0/1     Completed               3          26m
    kube-system     ingress-nginx-controller-7f9bbb757f-5st8d   1/1     Running                 0          26m
    kube-system     kube-apiserver-minikube                     1/1     Running                 0          31m
    kube-system     kube-controller-manager-minikube            1/1     Running                 0          31m
    kube-system     kube-proxy-ft5pv                            1/1     Running                 0          32m
    kube-system     kube-scheduler-minikube                     1/1     Running                 0          31m
    kube-system     storage-provisioner                         1/1     Running                 0          32m
    myworkload      myworkload-podA-0                           0/2     Init:ImagePullBackOff   0          16m
    myworkload      myworkload-podB-0                           0/1     ImagePullBackOff        0          16m
    myworkload      myworkload-podC-0                           1/1     Running                 0          16m
    myworkload      myworkload-podD-0                           0/1     ErrImagePull            0          16m
    myworkload      myworkload-podE-0                           1/1     Running                 0          16m
    myworkload      myworkload-podF-0                           0/1     Init:ImagePullBackOff   0          16m
    myworkload      myworkload-podG-0                           0/2     ErrImagePull            0          16m
    myworkload      myworkload-podH-0                           0/1     Running                 0          16m
    myworkload      myworkload-podI-0                           0/1     Running                 0          16m
    myworkload      myworkload-podJ-0                           0/1     Running                 0          16m
    myworkload      myworkload-podK-0                           0/1     ImagePullBackOff        0          16m
    myworkload      myworkload-podL-0                           0/1     ImagePullBackOff        0          16m

Describing the pods yields errors like this (hostnames and IP addresses obfuscated, newlines added for readability):

      Warning  Failed     103s     kubelet, minikube  Failed to pull image "<OUR_REGISTRY>:<OUR_REGISTRY_PORT>/mycompany/myimage:1.2.3.4": 
      rpc error: code = Unknown desc = Error response from daemon: Get http://<OUR_REGISTRY>:<OUR_REGISTRY_PORT>/v2/:
       dial tcp: lookup <OUR_REGISTRY> on 172.23.80.1:53: read udp 172.23.90.71:48824->172.23.80.1:53: i/o timeout

This is on Minikube 1.15.1
Minikube start command was

     minikube start --disk-size 40G --cpus 6 --insecure-registry <OUR_REGISTRY>:<OUR_REGISTRY_PORT> --kubernetes-version=v1.17.16 --memory 32g --mount --mount-string="C:\Users\<MY_USER>\minikube:/c/Users/<MY_USER>/minikube

Once the minikube VM is in this state, when I try to manually pull an image with the Docker CLI after setting minikube's docker-env, this gives me the same error. Further, virtually ANY DNS lookups from within the VM fail after that.

Interestingly, while I had a few issues here and there after changing my machine and switching from VirtualBox to Hyper-V, this worked for some time without too much problems. But now it happens consistently.

A few more things I did:

- I added an entry into my Windows hosts file for our internal registry - different from VirtualBox, Hyper-V VMs are not aware of our domain name and our default config will use our registry's non-qualified hostname, which wouldn't work. 
- After "minikube start", for my own convenience I added an entry for the current IP of the minikube VM into my Windows hosts file. Removing it doesn't seem to change the behavior, though.

I also googled for this issue, and found a few that appear similar, but mostly were reported in the context of a plain Docker4Windows installation (i.e., nothing related to minikube), e.g.

https://forums.docker.com/t/docker-login-dial-tcp-lookup-index-docker-io-read-udp-i-o-timeout/14844

Interestingly, after I had minikube working for a while, I installed Docker4Windows since this seemed the best way to get the Docker CLI on that machine... I am not a 100% sure, but I think the issues might have started around that time.

Any hints on how to further analyze the root cause of this issue welcome.

BTW: I also tried to switch back to using VirtualBox (which to my surprise actually seemed to work after skipping the check for active Hyper-V with the  
    
    --no-vtx-check=false

switch. Virtually same result (maybe a tad worse, with less successful pulls before everything falls apart).

**Comment by medyagh** [comment]
@jgoeres do u mind providing

`kubectl get pods -A`
and also the logs for the coreDNS pods in kube-system namspace ?

and also the full output of

```minikube logs```

/triage needs-information
/kind support

**Comment by jgoeres** [comment]
@medyagh Thanks for looking into this.

The sample output above is an (obfuscated, I just renamed my workload pods) output of kubectl get pods -A, I left it running over night, now it looks like this:

    $ kubectl get pods -A
    NAMESPACE       NAME                                        READY   STATUS                  RESTARTS   AGE
    kube-system     coredns-6955765f44-tbgrg                    1/1     Running                 0          13h
    kube-system     etcd-minikube                               1/1     Running                 0          13h
    kube-system     ingress-nginx-admission-create-dgl4c        0/1     Completed               0          13h
    kube-system     ingress-nginx-admission-patch-z9p96         0/1     Completed               3          13h
    kube-system     ingress-nginx-controller-7f9bbb757f-5st8d   1/1     Running                 0          13h
    kube-system     kube-apiserver-minikube                     1/1     Running                 0          13h
    kube-system     kube-controller-manager-minikube            1/1     Running                 0          13h
    kube-system     kube-proxy-ft5pv                            1/1     Running                 0          13h
    kube-system     kube-scheduler-minikube                     1/1     Running                 0          13h
    kube-system     storage-provisioner                         1/1     Running                 0          13h
    sparkoperator   sparkoperator-c8fd56c74-9zfxw               1/1     Running                 0          12h
    sparkoperator   sparkoperator-webhook-init-gq6hw            0/1     Completed               0          12h
    myworkload      myworkload-podA-0                           0/2     Init:ImagePullBackOff   0          12h
    myworkload      myworkload-podB-0                           0/1     ImagePullBackOff        0          12h
    myworkload      myworkload-podC-0                           1/1     Running                 0          12h
    myworkload      myworkload-podD-0                           0/1     ImagePullBackOff        0          12h
    myworkload      myworkload-podE-0                           1/1     Running                 0          12h
    myworkload      myworkload-podF-0                           0/1     Init:ImagePullBackOff   0          12h
    myworkload      myworkload-podG-0                           0/2     ImagePullBackOff        0          12h
    myworkload      myworkload-podH-0                           0/1     Running                 0          12h
    myworkload      myworkload-podI-0                           0/1     Running                 0          12h
    myworkload      myworkload-podJ-0                           0/1     Running                 0          12h
    myworkload      myworkload-podK-0                           0/1     ErrImagePull            0          12h
    myworkload      myworkload-podL-0                           0/1     ImagePullBackOff        0          12h

Minikube logs:

[minikube_logs_obfuscated.txt](https://github.com/kubernetes/minikube/files/5885251/minikube_logs_obfuscated.txt)

CoreDNS logs:

[coreDNS_logs.txt](https://github.com/kubernetes/minikube/files/5885256/coreDNS_logs.txt)

**Comment by jgoeres** [comment]
Some minor updates: I had two colleagues with almost the same machine setup as my own try this out. They are not affected. However, Minikube seems slow in their environments (and in my own, as far as it works at all). 
Interestingly, the problem worsened over the course of yesterday - where initially the pull of the images for the ingress controller addon would work most of the time, already this step failed (same DNS i/o timeout) when I tried later during the day.
I had left the machine stuck in that state (ingress controller images in "ErrImagePull"/"IMagePullBackOff" state over the night. 
This morning, the ingress controller images were there and I could deploy my workload and an additional operator just fine. 

I am currently suspecting a threat protection software that is forcibly installed on our dev machines to be the culprit - alas, as a user on the machine there is no means to actually see what it is doing.
Could it be that there is a lot of DNS resolution going on during image pull, which could be considered malicious by that software? I am probably the most intense user of the minikube environment in our unit, regularly throwing away and redeploying workloads, so my machine might have been the first to trigger the malware detection.

Still, if you have any other things to check or try out, please let me know.

**Comment by jgoeres** [comment]
When the minikube is in that state and I do a docker pull (either after "minikube" ssh or after using "minikube docker-env"), this will fail with the exact same error message (the message shown in "kubectl describe pod" is apparently just the error message passed through from Docker Daemon).
Interestingly, when the VM is "nearing" that state and I do a docker pull I also notice that download speed is extremely slow (<1MB/s), compared to when doing a pull with the Docker daemon installed via Docker Desktop), performance is fine (>25MB/s).
I meanwhile got access to the "logs" of the threat protection software - alas, it says nothing about blocking anything. 
That, and the fact that it happens regardless of the chosen hypervisor (I tried both Hyper-V and Virtualbox, and both are seeing the issue) points more to an internal problem of the minikube VM.

**Comment by jgoeres** [comment]
Okay, so I finally found the root cause for the observed behavior:
In the workload that I deployed, there was a job that amongst other things did a "curl" command. A change in its parametrization triggered a bug in the job script that would cause curl to try to curl from several non-existing hostnames. We have our own retry logic in that job (since Kubernetes retries have quite inflexible retry behaviour with long and quickly rising backoff intervals), so that the failed curl would be retried once every second. The container running that job is one of the smaller ones, so it would usually be among the first to actually reach running state, which explains why only some of the other pods would reach running (all those were the DNS lookup had already been done), while most would not. 

To reproduce the issue, I created a simple script and docker image:

**killDns.sh**

    #!/bin/bash
    
    if [[ $1 == "kill" ]]; then
      while [[ true ]]; do
         curl foo bar givesitnot
         sleep 1
      done
    elif [[ $1 == "check" ]]; then
    
      while [[ true ]]; do
         nslookup google.de | ts
         sleep 1
      done
    else
      echo "Bad command, use \"kill\" to try to kill DNS, use  \"check\" to test if lookups still work."
      exit 1
    fi


**Dockerfile**

    FROM alpine:3.12.3
    ADD killDns.sh /
    RUN apk update && \
        apk add \
            curl \
            bash \
            moreutils
    ENTRYPOINT [ "/killDns.sh"]
    CMD [ "check" ]

Put these two files in an otherwise empty dir. Open your favorite shell in that directory and use "minikube docker-env" to let your docker CLI talk to the docker daemon in the minikube VM, e.g., on bash shell, run 

    eval $(minikube docker-env)

Build it with

    docker build -t dnskiller:0.0.1 .

Now open another shell (and do minikube docker-env in it)
and run the image as plain docker container like this

    docker run --rm --name dnschecker -it dnskiller:0.0.1 -- check

This will give you the output of a suuccessful nslookup for an existing hostname every second:


    Feb 11 08:10:32 Server:         10.0.2.3
    Feb 11 08:10:32 Address:        10.0.2.3:53
    Feb 11 08:10:32
    Feb 11 08:10:32 Non-authoritative answer:
    Feb 11 08:10:32 Name:   google.de
    Feb 11 08:10:32 Address: 173.194.73.94
    Feb 11 08:10:32
    [...]

No run the image as plain docker container like this:

    docker run --rm --name dnskiller -it dnskiller:0.0.1 -- kill

This will now do the curl command against a bunch of non-existing hostnames - also once every second:

    $ docker run --rm --name dnskiller -it dnskiller:0.0.1 kill
    curl: (6) Could not resolve host: foo
    curl: (6) Could not resolve host: bar
    curl: (6) Could not resolve host: givesitnot
    curl: (6) Could not resolve host: foo
    [...]

Leave both containers running for a while and keep an eye on the output of the dnschecker container... and you will see that both continue to run just fine, the "dnskiller" with failing curl commands, the "dnschecker" with _successful_ nslookups.

Stop the dnskiller container. 
Next, in another shell, follow the logs of the coreDNS pod in minikube with 

    kubectl logs -n kube-system coredns-<THE_NAME_OF_YOUR_CORE_DNS_POD> -f


Now run the "dnskiller" again, but this time inside Kubernetes:

    kubectl run dnskiller -it --image dnskiller:0.0.1 -- kill

Again, keep an eye on the output of the dnschecker container - it will only take a few seconds to give you 

    Feb 10 09:56:39 ;; connection timed out; no servers could be reached

At the same moment, CoreDNS will start logging error messages:

    [ERROR] plugin/errors: 2 foo. A: read udp 172.17.0.2:59806->10.0.2.3:53: i/o timeout
    [ERROR] plugin/errors: 2 foo. A: read udp 172.17.0.2:39949->10.0.2.3:53: i/o timeout
    [ERROR] plugin/errors: 2 bar. AAAA: read udp 172.17.0.2:48435->10.0.2.3:53: i/o timeout
    [ERROR] plugin/errors: 2 bar. A: read udp 172.17.0.2:51295->10.0.2.3:53: i/o timeout
    [...]
    [ERROR] plugin/errors: 2 givesitnot. AAAA: read udp 172.17.0.2:52102->10.0.2.3:53: i/o timeout
    [ERROR] plugin/errors: 2 givesitnot. A: read udp 172.17.0.2:58473->10.0.2.3:53: i/o timeout
    [ERROR] plugin/errors: 2 givesitnot. AAAA: read udp 172.17.0.2:39532->10.0.2.3:53: i/o timeout
    [ERROR] plugin/errors: 2 givesitnot. A: read udp 172.17.0.2:35336->10.0.2.3:53: i/o timeout
    [ERROR] plugin/errors: 2 givesitnot. A: read udp 172.17.0.2:38487->10.0.2.3:53: i/o timeout
    [ERROR] plugin/errors: 2 givesitnot. AAAA: read udp 172.17.0.2:51494->10.0.2.3:53: i/o timeout
    [ERROR] plugin/errors: 2 givesitnot. AAAA: read udp 172.17.0.2:47124->10.0.2.3:53: i/o timeout
    [...]
    [ERROR] plugin/errors: 2 foo. A: read udp 172.17.0.2:59186->10.0.2.3:53: i/o timeout
    [ERROR] plugin/errors: 2 foo. AAAA: read udp 172.17.0.2:56560->10.0.2.3:53: i/o timeout
    [ERROR] plugin/errors: 2 foo. A: read udp 172.17.0.2:42531->10.0.2.3:53: i/o timeout
    [ERROR] plugin/errors: 2 foo. AAAA: read udp 172.17.0.2:42847->10.0.2.3:53: i/o timeout
    [...]

Stopping the "dnskiller" pod will eventually stop these log messages. However, any DNS lookups (like those done by the "dnschecker") will continue failing .

Funny enough, once I have the system in this state, it will not recover easily.
Even doing a "minikube delete" followed by a "minikube start" does not fix it. What DID fix it for me (aside from rebooting the Windows host) was deleting and starting minikube and also restarting the Windows "Host Network Service".
This also explains why I sometimes observed this issue already when enabling the ingress addon - since these are the first images that are actually pulled (where the K8s infra pods come preloaded with the image), this is the first time that DNS lookups are done.
I tried this with both Hyper-V and Virtualbox drivers, same result. 
Summary: So failing DNS requests appear to somehow derail CoreDNS, which itself does something bad with the upstream DNS server, which in the case of a Windows host causes permanent DNS failures. Would be interesting to see if the same happens in other environments (Minikube on Linux, EKS, AKS, etc.). So the question is: who is to blame? curl? CoreDNS? The Windows DNS service? Since doing the test without CoreDNS in the loop does not show the problem, it probably makes sense to start looking this way. But note that I did **not** see the issue when using the Docker driver with my "Docker Desktop" installation.

**Comment by priyawadhwa** [comment]
@jgoeres impressive detective work! 

I'm not sure what more we can do at this point, but if you find anything in minikube that can be changed to improve this we'd be happy to accept PRs. I'm going to leave your temporary solution here in case anybody else faces the same issue:

>What DID fix it for me (aside from rebooting the Windows host) was deleting and starting minikube and also restarting the Windows "Host Network Service".

**Comment by jgoeres** [comment]
At this point, I think that the issue would have to be further investigated on CoreDNS side (or maybe the Minikube network?), alas, the issue I reported with the CoreDNS project has been closed, since "we see this all time, k8s networking isn't reliable". Luckily, this isn't critical as it can be avoided and I also didn't manage to reproduce it on production environments (EKS, mostly).
So fine for me. :-)
